File: ios-sdk/Sources/SecuritySDK/CryptoEngine.swift
import CryptoKit
import Foundation

/// Native-only primitives mirroring Android `CryptoEngine` (no JavaScript crypto).
public enum CryptoEngine {
    public static func encryptAesGcm256(key: Data, plaintext: Data) throws -> (iv: Data, ciphertext: Data, tag: Data) {
        guard key.count == 32 else {
            throw NSError(domain: "CryptoEngine", code: 1, userInfo: [NSLocalizedDescriptionKey: "AES-256 key must be 32 bytes"])
        }
        let sym = SymmetricKey(data: key)
        let sealed = try AES.GCM.seal(plaintext, using: sym)
        let iv = Data(sealed.nonce)
        guard iv.count == 12 else {
            throw NSError(domain: "CryptoEngine", code: 2, userInfo: [NSLocalizedDescriptionKey: "unexpected IV length"])
        }
        return (iv, sealed.ciphertext, sealed.tag)
    }

    /// Decrypts and authenticates; throws on bad tag (CryptoKit authentication failure).
    public static func decryptAesGcm256(key: Data, iv: Data, ciphertext: Data, tag: Data) throws -> Data {
        guard key.count == 32 else {
            throw NSError(domain: "CryptoEngine", code: 1, userInfo: [NSLocalizedDescriptionKey: "AES-256 key must be 32 bytes"])
        }
        guard iv.count == 12, tag.count == 16 else {
            throw NSError(domain: "CryptoEngine", code: 3, userInfo: [NSLocalizedDescriptionKey: "invalid IV or tag length"])
        }
        let sym = SymmetricKey(data: key)
        let nonce = try AES.GCM.Nonce(data: iv)
        let box = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
        return try AES.GCM.open(box, using: sym)
    }

    public static func hkdfSha256(ikm: Data, salt: Data?, info: String, length: Int) -> Data {
        HkdfSha256.derive(ikm: ikm, salt: salt, info: info, length: length)
    }

    /// Verifies ECDSA P-256 (SHA-256) over `data` using peer SPKI DER.
    public static func verifyEcdsaSha256(data: Data, signatureDer: Data, peerPublicKeySpki: Data) -> Bool {
        guard let pub = try? P256.Signing.PublicKey(derRepresentation: peerPublicKeySpki) else {
            return false
        }
        guard let sig = try? P256.Signing.ECDSASignature(derRepresentation: signatureDer) else {
            return false
        }
        return pub.isValidSignature(sig, for: data)
    }
}

File: android-sdk/security-sdk/src/main/kotlin/com/bankingsdk/security/CryptoEngine.kt
package com.bankingsdk.security

import com.bankingsdk.security.crypto.AesGcm256
import com.bankingsdk.security.crypto.AesGcmCiphertext
import com.bankingsdk.security.crypto.EcP256
import com.bankingsdk.security.crypto.HkdfSha256
import java.security.Signature

/**
 * Native-only cryptographic primitives exposed for advanced integrations (still no JS crypto).
 *
 * AES-GCM: 12-byte random IV, 128-bit tag; decryption verifies the tag (AEAD) and fails on tampering.
 * ECC: NIST P-256 (secp256r1) for ECDH and ECDSA (SHA-256).
 */
object CryptoEngine {
    fun encryptAesGcm256(key: ByteArray, plaintext: ByteArray): AesGcmCiphertext =
        AesGcm256.encrypt(key, plaintext)

    /** Decrypts and authenticates; throws [javax.crypto.AEADBadTagException] if the tag does not match. */
    fun decryptAesGcm256(key: ByteArray, blob: AesGcmCiphertext): ByteArray =
        AesGcm256.decrypt(key, blob)

    fun hkdfSha256(ikm: ByteArray, salt: ByteArray?, info: String, length: Int): ByteArray =
        HkdfSha256.derive(ikm, salt, info, length)

    /**
     * Verify ECDSA P-256 signature (SHA-256) against an SPKI-encoded peer public key.
     */
    fun verifyEcdsaSha256(data: ByteArray, signatureDer: ByteArray, peerPublicKeySpki: ByteArray): Boolean {
        return try {
            val pub = EcP256.publicKeyFromSpki(peerPublicKeySpki)
            val sig = Signature.getInstance("SHA256withECDSA")
            sig.initVerify(pub)
            sig.update(data)
            sig.verify(signatureDer)
        } catch (_: Exception) {
            false
        }
    }
}
